azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,342 @@
+        cmd,
+        hub: Optional[str] = None,
+        rg: Optional[str] = None,
+        login: Optional[str] = None,
+        auth_type_dataplane: Optional[str] = None,
+    ):
+        super(StateProvider, self).__init__(
+            cmd=cmd,
+            hub_name=hub,
+            rg=rg,
+            login=login,
+            auth_type_dataplane=auth_type_dataplane
+        )
+        self.auth_type = auth_type_dataplane
+
+        # the login argument doesn't work later if auth type = "login" and a connection string isn't given initially
+        if login or auth_type_dataplane == "key":
+            self.login = self.target["cs"]
+        else:
+            self.login = None
+
+    def save_state(self, filename: str):
+        '''
+        Writes all hub configurations, device identities and device twins from the origin hub to a json file
+        '''
+
+        configs = iot_hub_configuration_list(cmd=self.cmd, hub_name=self.hub_name, resource_group_name=self.rg,
+                                             login=self.login, auth_type_dataplane=self.auth_type)
+
+        identities = iot_device_list(cmd=self.cmd, hub_name=self.hub_name, top=-1, resource_group_name=self.rg,
+                                     login=self.login, auth_type_dataplane=self.auth_type)
+
+        with open(filename, 'w', encoding='utf-8') as f:
+
+            json.dump(configs, f)
+            f.write('\n')
+
+            for i in tqdm(range(len(identities)), desc="Exporting devices"):
+                id = identities[i]
+
+                module_objs = iot_device_module_list(cmd=self.cmd, device_id=id["deviceId"], hub_name=self.hub_name,
+                                                     resource_group_name=self.rg, login=self.login,
+                                                     auth_type_dataplane=self.auth_type)
+
+                # number of modules in the device (not including the two modules automatically included in edge devices)
+                if(id["capabilities"]["iotEdge"]):
+                    id["numModules"] = len(module_objs) - 2
+                else:
+                    id["numModules"] = len(module_objs)
+
+                # primary and secondary keys show up in the "show" output but not in the "list" output
+                if id["authenticationType"] == DeviceAuthApiType.sas.value:
+                    id2 = _iot_device_show(self.target, id["deviceId"])
+                    id["symmetricKey"] = id2["authentication"]["symmetricKey"]
+
+                json.dump(id, f)
+                f.write('\n')
+
+                # this may be redundant, since it seems like the same fields are included in the identity
+                twin = _iot_device_twin_show(self.target, id["deviceId"])
+                json.dump(twin, f)
+                f.write('\n')
+
+                for module in module_objs:
+                    module = vars(module)
+
+                    if(module["module_id"] not in ["$edgeAgent", "$edgeHub"]):
```

Comment:
rework your if statements to be more pythonic:
if module["module_id"] not in ["$edgeAgent", "$edgeHub"]: